test-suite/shared/src/test/scala/org/scalajs/testsuite/niocharset/BaseCharsetTest.scala

```diff
@@ -30,6 +30,57 @@ class BaseCharsetTest(val charset: Charset) {
   protected def testDecode(in: ByteBuffer)(
       outParts: OutPart[CharBuffer]*): Unit = {
 
+    def testDecodeVsSteppedDecode(malformedAction: CodingErrorAction,
+        unmappableAction: CodingErrorAction) = {
+
+      val in1 = in.duplicate
+      val in2 = in.duplicate
```

Comment:
You should take `readOnly` into account, the same way it is done [in `testOneConfig`](https://github.com/scala-js/scala-js/blob/ed20d0dac5d92f4a84edf5837054bae70aecb021/test-suite/shared/src/test/scala/org/scalajs/testsuite/niocharset/BaseCharsetTest.scala#L40-L44) (so you should probably abstract that away in a `def duplicateIn(in, readOnly)` in order not to repeat it 3 times).

This is important to check that both code paths with `hasArray == true` and `hasArray == false` are handled by the charset decoders. In particular, for UTF-8, this means `decodeLoopArray` and `decodeLoopNoArray`.